### DISCUSSION TITLE
Discord Bot (Auto-rename when a user connects to a voice channel)

### DISCUSSION THREAD

**Discussion by Potatis04** [Rapptz/discord.py#9482] [Q&A]
Hello, I am currently trying to code a Discord bot so basically with a queue system; 

"Player xyz" joins a waiting room voice channel -> His username changes to "1) Player xyz", while he is waiting another user "Player abc" joins the voice chat and his username changes to "2) Player abc". Then each 10 or 5 minutes the names goes back to normal if they haven't left the voice channel.

I've asked ChatGPT to help me write some codes but I can't get them to work:
This is the most current code:
```py
print('
```import discord
from discord.ext import commands

intents = discord.Intents.default()
intents.voice_states = True

bot = commands.Bot(command_prefix="!", intents=intents)

queue_number = 1
voice_channel_id = 123456789  # Replace with the specific voice channel ID

@bot.event
async def on_voice_state_update(member, before, after):
    if after.channel is not None and after.channel.id == voice_channel_id:
        global queue_number
        await member.edit(nick=f"User {queue_number}")
        await member.send(f"You have joined the queue as User {queue_number}")
        queue_number += 1

bot.run("YOUR_BOT_TOKEN")
')```


######################################################
And this is the code (also from ChatGPT) before where I imported the bot from Youtube: https://www.youtube.com/watch?v=npyXUPX1a6Y&ab_channel=cdnAshton

```py
print('
import discord
from discord.ext import commands, tasks
import asyncio
import json

with open('config/config.json') as f:
    config = json.load(f)
    token = config['token']
    prefix = config['prefix']

intents = discord.Intents()
intents.typing = False
intents.presences = False

bot = commands.Bot(command_prefix=prefix, intents=intents)

queue = []
counter = 1

///# Create a dictionary to store the member's connected voice channel
connected_channels = {}

@bot.event
async def on_ready():
    print(f'Bot is ready!')

@bot.event
async def on_voice_state_update(member, before, after):
    global queue
    global connected_channels

    if after.channel:
        if after.channel.id == 1124832913695965214:  # Replace with the waiting room voice channel ID
            if member not in queue and member.id not in connected_channels:
                position = len(queue) + 1  # Calculate the position in the queue
                await member.edit(nick=f'{position}) Player {member.name}')
                connected_channels[member.id] = after.channel.id
    else:
        if member in queue:
            queue.remove(member)
            await member.edit(nick=member.name)
            if member.id in connected_channels:
                del connected_channels[member.id]


@tasks.loop(minutes=1)
async def check_change_names():
    global queue
    global connected_channels

    for member_id, channel_id in connected_channels.items():
        member = bot.get_user(member_id)
        channel = bot.get_channel(channel_id)

        if member and channel and member.voice is None:
            await member.edit(nick=member.name)
            del connected_channels[member_id]

@check_change_names.before_loop
async def before_check_change_names():
    await bot.wait_until_ready()
    await asyncio.sleep(3)  # Wait for 3 seconds before starting the task to allow the bot to fully connect

check_change_names.start()

@bot.command()
async def ping(ctx):
    await ctx.send("Pong!")

bot.run(token)
```')

**Comment by LostLuma** [comment]
Hello.

If you'd like to receive help with this issue you probably want to formulate your post in a more constructive manner, as you've not stated what issue you are running into besides that it "does not work", and also post the entirety of your code in code blocks rather than just subsections of it.

You can create code blocks like this:
\```py
print('Your code goes here.')
\```
It will turn into a more easily readable block that keeps indents:
```py
print('Your code goes here.')
```

You can also (continue) asking on the [official Discord server](https://discord.gg/r3sSKJJ) instead if you prefer, code blocks also work there.

___

Besides that you're probably also better off trying to learn programming on your own rather than using imitative AI tools such as ChatGPT, as it often produces misleading output and learning on your own will lead to more understanding of your own code and answers from others. There are excellent resources like the [official Python tutorial](https://docs.python.org/3/tutorial/index.html) (which is admittedly a bit long) and then [discord.py's own Getting started](https://discordpy.readthedocs.io/en/latest/#getting-started) section.

**Reply by Potatis04** [reply]
Hi and thank you for your response! I am really new to this platform and I am unfortunately not familiar with how to post.
The part that doesn't work is just the Bot's functionality, I've started it and it shows the bot as online but no functions work, not even the "!ping" command. I was just hoping for some clearance as this is most likely a one time project

**Reply by No767** [reply]
In response to your question, you will need to have the intent `message_content` enabled. Recently (2022) Discord requires that intent to be enabled. More info about intents can be found [here](https://discordpy.readthedocs.io/en/latest/intents.html#privileged-intents)

**Reply by LostLuma** [reply]
Thanks for updating the post.

As mentioned, for the text command to work you will need to enable the message content intent, for the voice state update you need the voice states intent as [mentioned here](https://discordpy.readthedocs.io/en/latest/api.html#discord.on_voice_state_update). To fill in other data like the `channel` field on the voice state you also need the `guilds` intent, otherwise the `if after.channel:` check will always fail, regardless of whether the user did really join a channel.

The only other issue I see at a quick glance is that you're trying to edit a `User` instead of a `Member` in the `check_change_names` task. A User is a person that is not attached to a certain guild, useful for example if you're sending them a DM. To edit their nick you need a Member which is part of a guild.

To achieve this the bot has a `get_guild` method that takes the ID of the guild, which in turn has a `get_member` method for getting the member which you can then edit the nick on.

**Reply by Potatis04** [reply]
This is the most recent piece of code I've updated and I still cannot get it to work. I have added an message intent and changed the "user" to "member". I have also enabled the guild and message intents on the Discord Developer website.

```
import discord
from discord.ext import commands, tasks
import asyncio
import nest_asyncio

token = ''TOKEN"
prefix = '!'

intents = discord.Intents()
intents.typing = False
intents.presences = False
intents.message_content = True

bot = commands.Bot(command_prefix=prefix, intents=intents)

queue = []
counter = 1

connected_channels = {}

@bot.event
async def on_ready():
    print(f'POTATOOOOOOOO!')

@bot.event
async def on_voice_state_update(member, before, after):
    global queue
    global connected_channels

    if after.channel:
        if after.channel.id == 1124832913695965214:  # Replace with the waiting room voice channel ID
            if member not in queue and member.id not in connected_channels:
                position = len(queue) + 1
                queue.append(member)
                await member.edit(nick=f'{position}) Player {member.name}')
                connected_channels[member.id] = after.channel.id
    else:
        if member in queue:
            queue.remove(member)
            await member.edit(nick=member.name)
            if member.id in connected_channels:
                del connected_channels[member.id]

@tasks.loop(seconds=5)
async def check_change_names():
    global queue
    global connected_channels

    for member_id, channel_id in connected_channels.items():
        member = bot.get_guild(1124832913242992751).get_member(member_id)
        channel = bot.get_channel(channel_id)

        if member and channel and member.voice is None:
            await member.edit(nick=member.name)
            del connected_channels[member_id]
        elif member and channel and member.voice.channel != channel:
            if member_id in connected_channels:
                time_since_last_update = (discord.utils.utcnow() - member.nick_updated_at).total_seconds()
                if time_since_last_update >= 300:  # 5 minutes (300 seconds)
                    await member.edit(nick=member.name)
                    del connected_channels[member_id]

@check_change_names.before_loop
async def before_check_change_names():
    await bot.wait_until_ready()
    await asyncio.sleep(3)

@bot.command()
async def ping(ctx):
    await ctx.send('Pong!')

async def run_bot():
    await bot.start(token)
    bot.loop.create_task(check_change_names.start())
    try:
        await bot.wait_until_ready()
        bot.loop.run_forever()
    except KeyboardInterrupt:
        bot.loop.run_until_complete(bot.logout())
    finally:
        bot.loop.close()

if __name__ == "__main__":
    asyncio.run(run_bot())
```

**Reply by ClementJ18** [reply]
Reset your token

**Reply by Potatis04** [reply]
I've done that, the reason the code says "TOKEN" is because Discord and Github doesn't let anyone post a Token ID without them spamming you

**Reply by LostLuma** [reply]
>  I have also enabled the guild and message intents on the Discord Developer website.

You need to enable the three intents in your code, not on the website, otherwise it will not do anything.

**Reply by Potatis04** [reply]
I've made the changes you mentioned and the bot runs normally by saying "POTATOOOOOOOOO" in the cmd, however none of the functions work, I've enabled the intents in the code and on the website.

```
import discord
from discord.ext import commands, tasks
import asyncio
import nest_asyncio

token = 'TOKEN (i have changed the token because of restrictions, when i put it back i will re-enter the token)'
prefix = '!'

intents = discord.Intents()
intents.typing = True
intents.presences = True
intents.messages = True

bot = commands.Bot(command_prefix=prefix, intents=intents)

queue = []
counter = 1

connected_channels = {}

@bot.event
async def on_ready():
    print(f'POTATOOOOOOOO!')

@bot.event
async def on_voice_state_update(member, before, after):
    global queue
    global connected_channels

    if after.channel:
        if after.channel.id == 1124832913695965214:  # Replace with the waiting room voice channel ID
            if member not in queue and member.id not in connected_channels:
                position = len(queue) + 1
                queue.append(member)
                await member.edit(nick=f'{position}) Player {member.name}')
                connected_channels[member.id] = after.channel.id
    else:
        if member in queue:
            queue.remove(member)
            await member.edit(nick=member.name)
            if member.id in connected_channels:
                del connected_channels[member.id]

@tasks.loop(seconds=5)
async def check_change_names():
    global queue
    global connected_channels

    for member_id, channel_id in connected_channels.items():
        member = bot.get_guild(1124832913242992751).get_member(member_id)
        channel = bot.get_channel(channel_id)

        if member and channel and member.voice is None:
            await member.edit(nick=member.name)
            del connected_channels[member_id]
        elif member and channel and member.voice.channel != channel:
            if member_id in connected_channels:
                time_since_last_update = (discord.utils.utcnow() - member.nick_updated_at).total_seconds()
                if time_since_last_update >= 300:  # 5 minutes (300 seconds)
                    await member.edit(nick=member.name)
                    del connected_channels[member_id]

@check_change_names.before_loop
async def before_check_change_names():
    await bot.wait_until_ready()
    await asyncio.sleep(3)

@bot.command()
async def ping(ctx):
    await ctx.send('Pong!')

async def run_bot():
    await bot.start(token)
    bot.loop.create_task(check_change_names.start())
    try:
        await bot.wait_until_ready()
        bot.loop.run_forever()
    except KeyboardInterrupt:
        bot.loop.run_until_complete(bot.logout())
    finally:
        bot.loop.close()

if __name__ == "__main__":
    asyncio.run(run_bot())
    ```

**Reply by LostLuma** [reply]
You're stilling missing the guilds, voice_states, and messages intents.

**Reply by Potatis04** [reply]
Could you please write the intents I have to write in the code? I am very new to this, sorry

**Reply by LostLuma** [reply]
The names are the exact ones I used in the message. You add them like you did for example the typing intent.

**Reply by Potatis04** [reply]
so like this?

```
intents = discord.Intents()
intents.typing = True
intents.presences = True
intents.messages = True
intents.guilds = True
intents.voice_states = True
intents.messages_intents = True
```

**Reply by LostLuma** [reply]
Yes.

**Reply by Potatis04** [reply]
it gives me this response when i try to run the bot:

Traceback (most recent call last):
  File "C:\Users\potatis\OneDrive\Desktop\NY\bot.py", line 15, in <module>
    intents.messages_intents = True
    ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Intents' object has no attribute 'messages_intents'

**Reply by itswilliboy** [reply]
It's `Intents.message_content`. See [the docs](https://discordpy.readthedocs.io/en/stable/api.html?highlight=intents#discord.Intents.message_content).

**Reply by Potatis04** [reply]
thanks, the !ping function now works, however the queue system for the voice channels is still not working as it is supposed to. For context I want a queue system for waiting rooms in discord where the first guy that connects to a voice channel gets his username changed from "Player abc" to "1) Player abc" and if the first guy is sitting there then if a second guy connects he gets "2) Player xyz". The number should be reset each 5 / 10 minutes if the member isn't connected to the specific vc and the names should change as soon as they connect to the vc (or with a 3-10 second delay)

This is the current full code:

```
import discord
from discord.ext import commands, tasks
import asyncio
import nest_asyncio

token = '(Removed for safety reason)'
prefix = '!'

intents = discord.Intents()
intents.typing = True
intents.presences = True
intents.messages = True
intents.guilds = True
intents.voice_states = True
intents.message_content = True

bot = commands.Bot(command_prefix=prefix, intents=intents)

queue = []
counter = 1

connected_channels = {}

@bot.event
async def on_ready():
    print(f'POTATOOOOOOOO!')

@bot.event
async def on_voice_state_update(member, before, after):
    global queue
    global connected_channels

    if after.channel:
        if after.channel.id == 1124832913695965214:  # Replace with the waiting room voice channel ID
            if member not in queue and member.id not in connected_channels:
                position = len(queue) + 1
                queue.append(member)
                await member.edit(nick=f'{position}) Player {member.name}')
                connected_channels[member.id] = after.channel.id
    else:
        if member in queue:
            queue.remove(member)
            await member.edit(nick=member.name)
            if member.id in connected_channels:
                del connected_channels[member.id]

@tasks.loop(seconds=5)
async def check_change_names():
    global queue
    global connected_channels

    for member_id, channel_id in connected_channels.items():
        member = bot.get_guild(1124832913242992751).get_member(member_id)
        channel = bot.get_channel(channel_id)

        if member and channel and member.voice is None:
            await member.edit(nick=member.name)
            del connected_channels[member_id]
        elif member and channel and member.voice.channel != channel:
            if member_id in connected_channels:
                time_since_last_update = (discord.utils.utcnow() - member.nick_updated_at).total_seconds()
                if time_since_last_update >= 300:  # 5 minutes (300 seconds)
                    await member.edit(nick=member.name)
                    del connected_channels[member_id]

@check_change_names.before_loop
async def before_check_change_names():
    await bot.wait_until_ready()
    await asyncio.sleep(3)

@bot.command()
async def ping(ctx):
    await ctx.send('Pong!')

async def run_bot():
    await bot.start(token)
    bot.loop.create_task(check_change_names.start())
    try:
        await bot.wait_until_ready()
        bot.loop.run_forever()
    except KeyboardInterrupt:
        bot.loop.run_until_complete(bot.logout())
    finally:
        bot.loop.close()

if __name__ == "__main__":
    asyncio.run(run_bot())
```

**Reply by LostLuma** [reply]
The basic system works fine for me as you describe it, using this exact code. However you've seemingly made up the `member.nick_updated_at` attribute so the task will fail to rename users back to their original names when they have switched channels.

Since I can not find any reference to similar code existing on the internet, you repeatedly demonstrating a lack of basic basic problem solving, and your last message sounding an awful lot like a prompt for ChatGPT I consider it quite rude that you are not trying to actually understand the code you're writing, but instead using help from here as a substitute when it can no longer imitate real knowledge and understanding.

As a result of this I am no longer interested in replying to this thread, good luck on your programming adventures.

**Reply by Potatis04** [reply]
Sorry if it came across that way, I'm just trying to understand where I've made the mistakes. I really just created my profile here and I really have no clue how people here operate. Sorry :(
